File: php/slim-doctrine/app/src/com/sprint/sms/api/controller/MenuController.php
<?php
namespace App\com\sprint\sms\api\controller;

use \App\com\sprint\sms\api\bean\Response;
use \App\com\sprint\sms\api\util\RequestUtil;
use \App\com\sprint\sms\api\util\ObjectUtil;
use \App\com\sprint\sms\api\cache\Constant;
use \App\com\sprint\sms\api\domain\Menu;
use \App\com\sprint\sms\api\dao\MenuDao;
use \App\com\sprint\sms\api\base\BaseController;

class MenuController extends BaseController
{
	public function all() {
		$result = $this->getMenuDao()->getList();
		return Response::SUCCESS($result);
	}
	
	public function view() {
		$request = $this->getRequest();
		$id = RequestUtil::paramsToId($request);
		if (!isset($id)) {
			return Response::ERROR_CODE("01", "id is required");
		}
		$entity = $this->getMenuDao()->get($id);
		return Response::SUCCESS($entity);
	}
	
	public function create() {
		$request = $this->getRequest();
		$entity = RequestUtil::paramsToObject($request, Menu::class);
		$parentObj = $entity->getParent();
		$parentId = $parentObj !== null ? $parentObj->getId() : null;
		$parentId = $parentId !== null ? trim($parentId) : "";
		if ($parentId !== "") {
			$parentObj = $this->getMenuDao()->get($parentId);
			if ($parentObj === null) {
				return Response::ERROR_CODE("01", "Parent is not found");
			}
		} else {
			$parentObj = null;
		}
		$entity->setParent($parentObj);
		$entity = $this->getMenuDao()->save($entity);
		return Response::SUCCESS($entity); 
	}
	
	public function update() {
		$request = $this->getRequest();
		$new = RequestUtil::paramsToObject($request, Menu::class);
		$id = $new->getId();
		if (!isset($id)) {
			return Response::ERROR_CODE("01", "id is required");
		}
		$old = $this->getMenuDao()->get($id);
		if ($old == null) {
			return Response::ERROR_CODE("02", "Menu is not found");
		}
		$parentObj = $new->getParent();
		$parentId = $parentObj !== null ? $parentObj->getId() : null;
		if ($parentId !== null) {
			$parentId = $parentId !== null ? trim($parentId) : "";
			if ($parentId !== "") {
				if ($old->getParent() === null || $parentId !== $old->getParent()->getId()) {
					$parentObj = $this->getMenuDao()->get($parentId);
					if ($parentObj === null) {
						return Response::ERROR_CODE("03", "Parent is not found");
					}
					$old->setParent($parentObj);
				}
			} else {
				$old->setParent();
			}			
		}		
		$ignoredField = array("parent", "priority");
		$ignoredField = array_merge($ignoredField, ObjectUtil::getDefaultIgnoredField());
		ObjectUtil::copy($old, $new, $ignoredField, array(
			"title" => array(ObjectUtil::NOT_NULL, ObjectUtil::NOT_EMPTY),
			"link" => array(ObjectUtil::NOT_NULL),
			"icon" => array(ObjectUtil::NOT_NULL),
			"description" => array(ObjectUtil::NOT_NULL),
			"active" => array(ObjectUtil::NOT_NULL),
			"global" => array(ObjectUtil::NOT_NULL),
			"action" => array(ObjectUtil::NOT_NULL)
		));				
		$entity = $this->getMenuDao()->save($old);		
		return Response::SUCCESS($entity);
	}
	
	public function delete() {
		$request = $this->getRequest();
		$id = RequestUtil::paramsToId($request);
		if (!isset($id)) {
			return Response::ERROR_CODE("01", "id is required");
		}
		$entity = $this->getMenuDao()->delete($id);
		return Response::SUCCESS($entity);
	}
	
	public function sort() {
		$request = $this->getRequest();
		$id = RequestUtil::paramsToId($request);		
		$up = $request->getParam("up");
		$moveUp = "1" === $up || "true" === $up;
		$result = $this->getMenuDao()->sort($id, $moveUp);
		return Response::SUCCESS($result ? 1 : 0);
	}
	
	private function getMenuDao() {
		return new MenuDao($this->getEntityManager(), $this->getCache(), $this->getLogger());
	}
	
}

File: java/spring-hibernate/src/main/java/com/sprint/sms/api/controller/MenuController.java
package com.sprint.sms.api.controller;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

import javax.servlet.http.HttpServletRequest;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import com.ideahut.common2.dto.Response;
import com.sprint.sms.api.base.BaseController;
import com.sprint.sms.api.dao.MenuDao;
import com.sprint.sms.api.domain.Menu;
import com.sprint.sms.api.util.AppConstant;
import com.sprint.sms.api.util.ObjectUtil;
import com.sprint.sms.api.util.RequestUtil;

@Controller
@RequestMapping(value = "/menu", method = RequestMethod.POST)
public class MenuController extends BaseController {
	
	private static final Set<String> ignoredField;
	private static final Map<String, List<Integer>> rule;
	static {
		ignoredField = new HashSet<String>(ObjectUtil.getDefaultIgnoredField());
		ignoredField.add("parent");
		ignoredField.add("priority");
		rule = new TreeMap<String, List<Integer>>();
		rule.put("title", Arrays.asList(ObjectUtil.NOT_NULL, ObjectUtil.NOT_EMPTY));
		rule.put("link", Arrays.asList(ObjectUtil.NOT_NULL));
		rule.put("icon", Arrays.asList(ObjectUtil.NOT_NULL));
		rule.put("description", Arrays.asList(ObjectUtil.NOT_NULL));
		rule.put("active", Arrays.asList(ObjectUtil.NOT_NULL));
		rule.put("global", Arrays.asList(ObjectUtil.NOT_NULL));
		rule.put("action", Arrays.asList(ObjectUtil.NOT_NULL));
	}
	
	
	@Autowired
	private MenuDao menuDao;
	
	@RequestMapping("/all")
	@Transactional
	public Response all() {
		List<Menu> list = menuDao.getList();
		return Response.SUCCESS(list);
	}
	
	@RequestMapping("/view")
	@Transactional
	public Response view() {
		HttpServletRequest request = getRequest();
		Long id = RequestUtil.paramsToId(request, Long.class);
		if (id == null) {
			return Response.ERROR("01", "id is required");
		}
		Menu entity = menuDao.get(id);
		return Response.SUCCESS(entity);
	}
	
	@RequestMapping("/create")
	@Transactional
	public Response create() {
		HttpServletRequest request = getRequest();
		Menu entity = RequestUtil.paramsToObject(request, Menu.class);
		Menu parentObj = entity.getParent();
		Long parentId = parentObj != null ? parentObj.getId() : null;
		if (parentId != null) {
			parentObj = menuDao.get(parentId);
			if (parentObj == null) {
				return Response.ERROR("01", "Parent is not found");
			}
		} else {
			parentObj = null;
		}
		entity.setParent(parentObj);
		Long priority = menuDao.getMaxPriority();
		entity.setPriority(priority + 1);		
		entity = menuDao.save(entity);
		return Response.SUCCESS(entity);
	}
	
	@RequestMapping("/update")
	@Transactional
	public Response update() {
		HttpServletRequest request = getRequest();
		Menu newEntity = RequestUtil.paramsToObject(request, Menu.class);
		Long id = newEntity.getId();
		if (id == null) {
			return Response.ERROR("01", "id is required");
		}
		Menu oldEntity = menuDao.get(id);
		if (oldEntity == null) {
			return Response.ERROR("02", "Menu is not found");
		}
		String parentId = request.getParameter("parent" + AppConstant.SPLIT_OBJECT_FIELD + Menu.ID);
		if (parentId != null) {
			parentId = parentId.trim();
			if (parentId.length() != 0) {
				Long pid = new Long(parentId);
				if (oldEntity.getParent() == null || !pid.equals(oldEntity.getParent().getId())) {
					Menu parent = menuDao.get(pid);
					if (parent == null) {
						return Response.ERROR("03", "Parent is not found");
					}
					oldEntity.setParent(parent);
				}
			} else {
				oldEntity.setParent(null);
			}
		}
		oldEntity = ObjectUtil.copy(Menu.class, oldEntity, newEntity, ignoredField, rule);				
		Menu entity = menuDao.save(oldEntity);		
		return Response.SUCCESS(entity);
		
		
	}
	
	@RequestMapping("/delete")
	@Transactional
	public Response delete() {
		HttpServletRequest request = getRequest();
		Long id = RequestUtil.paramsToId(request, Long.class);
		if (id == null) {
			return Response.ERROR("01", "id is required");
		}
		Menu entity = menuDao.delete(id);
		return Response.SUCCESS(entity);
	}
	
	@RequestMapping("/sort")
	@Transactional
	public Response sort() {
		HttpServletRequest request = getRequest();
		Long id = RequestUtil.paramsToId(request, Long.class);
		String up = request.getParameter("up");
		boolean moveUp = "1".equals(up) || "true".equalsIgnoreCase(up);
		Boolean result = menuDao.sort(id, moveUp);
		return Response.SUCCESS(result);
	}
	
}
